crypto/s2n_tls13_keys.c

```diff
@@ -343,3 +344,41 @@ int s2n_tls13_update_application_traffic_secret(struct s2n_tls13_keys *keys, str
 
     return 0;
 }
+
+int s2n_tls13_derive_resumption_master_secret(struct s2n_tls13_keys *keys, struct s2n_hash_state *hashes, struct s2n_blob *secret_blob)
+{
+    notnull_check(keys);
+    notnull_check(hashes);
+    notnull_check(secret_blob);
+
+    /* Sanity check that input hash is of expected type */
+    S2N_ERROR_IF(keys->hash_algorithm != hashes->alg, S2N_ERR_HASH_INVALID_ALGORITHM);
+
+    s2n_tls13_key_blob(message_digest, keys->size);
+
+    /* copy the hashes into the message_digest */
+    DEFER_CLEANUP(struct s2n_hash_state hkdf_hash_copy, s2n_hash_free);
+    GUARD(s2n_hash_new(&hkdf_hash_copy));
+    GUARD(s2n_hash_copy(&hkdf_hash_copy, hashes));
+    GUARD(s2n_hash_digest(&hkdf_hash_copy, message_digest.data, message_digest.size));
+
+    /* Derive master session resumption from master secret */
+    GUARD(s2n_hkdf_expand_label(&keys->hmac, keys->hmac_algorithm, &keys->extract_secret,
+        &s2n_tls13_label_resumption_master_secret, &message_digest, secret_blob));
+
+    return S2N_SUCCESS;
+}
+
+S2N_RESULT s2n_tls13_derive_session_ticket_secret(struct s2n_tls13_keys *keys, struct s2n_blob *resumption_secret, struct s2n_blob *ticket_nonce, struct s2n_blob *secret_blob)
```

Comment:
Why is only this method S2N_RESULT?

Also, nit: this line is pretty long; you probably need to wrap

Reply:
The other methods call s2n_tls13_key_blob or s2n_tls13_connection_keys, both of which are macros that don't have the correct return value for S2N_RESULT.

Reply:
@maddeleine and I talked yesterday about this issue. I was able to come up with a possible solution but it would require a more general macro without all of the magic: https://godbolt.org/z/f437ae